### PR TITLE
Backfill remote event fetched by MSC3030  so we can paginate  from it later

### PR DESCRIPTION
**Pull request by MadLittleMods**
Backfill remote event fetched by MSC3030 `/timestamp_to_event` so we can paginate from it later

 - [x] ~~Depends on https://github.com/matrix-org/synapse/pull/13320~~

Complement tests: https://github.com/matrix-org/complement/pull/406

We could use the same method to backfill for `/context` as well in the future, see https://github.com/matrix-org/synapse/issues/3848

### Dev notes

```
COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestJumpToDateEndpoint
```

```
COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestJumpToDateEndpoint/parallel/federation/can_paginate_after_getting_remote_event_from_timestamp_to_event_endpoint
```

```
docker build -t matrixdotorg/synapse -f docker/Dockerfile .
```

#### Random snippets

For `asdf_get_debug_events_in_room_ordered_by_depth` to dump the events from the database in a room, see the dev notes in https://github.com/matrix-org/synapse/pull/12370. And `poetry add tabulate`

For a version without dependencies, see the dev notes for `asdf_get_debug_events_in_room_ordered_by_depth` in https://github.com/matrix-org/matrix-react-sdk/pull/8354

---

```py
logger.info("backfill_event event=%s", pformat(vars(event)))
logger.info("backfill_event eventAsdf=%s", pformat(vars(eventAsdf)))
```

```py
def get_debug_internal_metadata(ev: EventBase) -> dict[str, str]:
    dict = ev.internal_metadata.get_dict()
    return {
        "outlier": ev.internal_metadata.outlier,
        "stream_ordering": ev.internal_metadata.stream_ordering,
        "out_of_band_membership": dict.get("out_of_band_membership"),
        "send_on_behalf_of": dict.get("send_on_behalf_of"),
        "recheck_redaction": dict.get("recheck_redaction"),
        "soft_failed": dict.get("soft_failed"),
        "proactively_send": dict.get("proactively_send"),
        "redacted": dict.get("redacted"),
        "txn_id": dict.get("txn_id"),
        "token_id": dict.get("token_id"),
        "historical": dict.get("historical"),
    }

logger.info(
    "backfill_event event.internal_metadata=%s",
    pformat(get_debug_internal_metadata(event)),
)
logger.info(
    "backfill_event eventAsdf.internal_metadata=%s",
    pformat(get_debug_internal_metadata(eventAsdf)),
)
```

via https://stackoverflow.com/a/54653137/796832
```py
def exception_with_traceback(message):
    import sys
    import types

    tb = None
    depth = 0
    while True:
        try:
            frame = sys._getframe(depth)
            depth += 1
        except ValueError as exc:
            break

        tb = types.TracebackType(tb, frame, frame.f_lasti, frame.f_lineno)

    return Exception(message).with_traceback(tb)

asdf_exc = exception_with_traceback("asdf")
logger.error(
    "_update_outliers_txn event_id=%s outlier_persisted=%s",
    event.event_id,
    outlier_persisted,
    exc_info=(type(asdf_exc), asdf_exc, asdf_exc.__traceback__),
)
```

```py
import traceback

logger.info(
    "persist_events _event_persist_queue event_id=%s backfilled=%s stack=%s",
    event.event_id,
    backfilled,
    traceback.format_stack(),
)
```

#### Call stacks

```
persist_events or persist_event
_EventPeristenceQueue ->
_persist_event_batch
_persist_events_and_state_updates
_persist_events_txn
_update_outliers_txn ("Updating state for ex-outlier event")
```

```
_get_event_cache

_invalidate_get_event_cache
```

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/federation_event.py [hunk 1]
```diff
@@ -1315,6 +1320,42 @@ async def _handle_marker_event(self, origin: str, marker_event: EventBase) -> No
...
+        eventAsdf = await self._federation_client.get_pdu(
+            [destination],
+            event_id,
+            room_version,
+        )
+        event = await self._federation_client.get_pdu_from_destination_raw(
+            destination,
+            event_id,
+            room_version,
+        )
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1339]: What makes the event returned by `get_pdu_from_destination_raw(...)` work when processed by `_process_pulled_events` but not `get_pdu(...)`? By work, I mean is accepted and the test in https://github.com/matrix-org/complement/pull/406 passes

I know that `get_pdu_from_destination_raw(...)` returns a non-outlier but there is an additional difference because setting `event.internal_metadata.outlier = False` for `get_pdu(...)` still doesn't work completely.

File: synapse/handlers/federation_event.py [hunk 2]
```diff
@@ -765,6 +766,11 @@ async def _process_pulled_event(
         """
         logger.info("Processing pulled event %s", event)
 
+        # TODO: Why does this matter? The whole point of this function is to
+        # persist random PDU's from backfill. It shouldn't matter whether we saw
+        # them somewhere else first as an outlier, then during backfill. This
+        # function handles de-outliering anyway.
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:772]: It seems like the only blocker here is the `assert not outlier` in a few places and that we `compute_event_context` which only works for non-outliers. The silly part is that raw PDU's over federation aren't `outliers` which is why normal backfill just works. But using the cached getters sometimes gives us an `outlier`. Seems like we should just be converting every event to a non-outlier in `_process_pulled_events` 🤷 


```
_process_pulled_events
_process_pulled_event
_process_received_pdu
compute_event_context
```

File: synapse/handlers/federation_event.py [hunk 3]
```diff
@@ -1315,6 +1318,47 @@ async def _handle_marker_event(self, origin: str, marker_event: EventBase) -> No
...
+        # `_process_pulled_events()` -> `_update_outliers_txn()` to create a
+        # `state_group` and mimics what would happen in a regular backfill.
+        # `get_pdu()` can potentially return an `outlier` depending on the cache
+        # which we don't want.
+        event_non_outlier = make_event_from_dict(
+            event_from_response.get_pdu_json(),
+            event_from_response.room_version,
+            internal_metadata_dict=None,
+        )
+        assert not event_non_outlier.internal_metadata.outlier
```
Comment (MadLittleMods) [inline synapse/handlers/federation_event.py:1353]: See https://github.com/matrix-org/synapse/pull/13205#discussion_r918365746 for an alternative to handle this in the `_process_pulled_events` stack instead

File: synapse/handlers/federation_event.py (hunk 1 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: The difference is just the `outlier` vs non-`outlier`. We want a non-`outlier` so that `_process_pulled_events()` -> `_update_outliers_txn()` creates a `state_group` and mimics what would happen in a regular backfill. `get_pdu()` can potentially return an `outlier` depending on the cache which we don't want.

Things also didn't work before because I was directly editing the `event.internal_metadata.outlier` reference which is also referenced in other places like the cache. So I was unintentionally change the cache source of truth which threw other things off.

File: synapse/handlers/federation_event.py (hunk 2 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:771]: (I'm looking for your thoughts here)

File: synapse/handlers/room.py [hunk 4]
```diff
@@ -1450,25 +1451,37 @@ async def get_event_for_timestamp(
...
+                    # point.
+                    #
+                    # FIXME: After this backfill, we might want to run this
+                    # whole `get_event_for_timestamp` function again to make
+                    # sure they didn't give us an event from their gappy
+                    # history. Also need a heuristic for when to stop recursing
+                    # if they keep giving us gappy results.
+                    await self.federation_event_handler.backfill_event(
+                        domain, room_id, remote_event_id
+                    )
```
Comment (MadLittleMods) [inline synapse/handlers/room.py:1468]: Main part of the PR is here. We now backfill the event that we fetch from the remote federated server.

**Review by richvdh — COMMENTED** [review]
seems generally plausible! A few comments and thoughts. Generally it seems like you're hitting https://github.com/matrix-org/synapse/issues/12584 and we should probably set about fixing that first.

File: synapse/handlers/federation_event.py [hunk 5]
```diff
@@ -918,15 +923,12 @@ async def _get_state_ids_after_missing_prev_event(
         event_id: str,
     ) -> StateMap[str]:
         """Requests all of the room state at a given event from a remote homeserver.
-
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:921]: I wanna say auto-format on this one when I had logs in here. Sorry for not spotting it before handing it off for review. Removed ⏩

File: synapse/handlers/federation_event.py (hunk 2 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:772]: > > TODO: Why does this matter?
>
> well, this function performs a bunch of operations that are not appropriate when we persist an outlier (in particular, it makes sure we have all the `prev_events` and resolves the state across those prev events).

Ahh, so it's more about doing the expected operations for persisting an outlier rather than we can't handle it (because we can). A safe-guard in case someone in the future refactors Synapse to persist outliers using this function (i.e. they should use something else).

In our case, we're backfilling an event so this usage is appropriate.

In that case, this needs a good comment and a good assertion message. Something like: "This event is marked as an outlier. If you're trying to persist an outlier, you should use a different function. If you're trying to backfill an event, this is the right method but you need pass in an event copy that doesn't have `event.internal_metada.outlier = true`."

Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: Updated 👍 I'm sure some language changes are needed

https://github.com/matrix-org/synapse/blob/337d8be9ea5fe10a563f0293e466eb4b79403477/synapse/handlers/federation_event.py#L769-L784

File: synapse/handlers/federation_event.py (hunk 3 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: > maybe it should happen in get_pdu instead?

Good suggestion 👍

I've updated `get_pdu` to always return a copy of the event so that it's cache can stay pristine with the original PDU.

File: synapse/handlers/room.py [hunk 6]
```diff
@@ -1450,25 +1451,37 @@ async def get_event_for_timestamp(
...
 
+                    # Backfill this event so we can get a pagination token for
+                    # it with `/context` and paginate `/messages` from this
+                    # point.
+                    #
+                    # FIXME: After this backfill, we might want to run this
+                    # whole `get_event_for_timestamp` function again to make
+                    # sure they didn't give us an event from their gappy
+                    # history. Also need a heuristic for when to stop recursing
+                    # if they keep giving us gappy results.
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/room.py:1494]: On a client, `GET /timestamp_to_event?dir=b&ts=<the_ts>`, the ideal answer is `event4`

Actual events:
```
create_event
event1
event2
event3
event4
<ts corresponds to here>
event5
event6
event7
event8
event9
```


On our local server, we find the `create_event` locally next to a forward gap, so we go ask the remote homeserver.
Local server:
```
create_event
<gap>
event9
```

On the remote server we ask, they find `event2` is closest which is returned. But it's also next to a forward gap and not the true closest event.
Remote server:
```
create_event
event1
event2
<gap>
event9
```

So on our local server, after we backfill `event2`, we can see that it's still next to a forward gap. Currently, we just return the closest event (the local one or the remote one we asked about).

In the future, it might be good to be smarter about figuring out the gap to see if we can determine the true closest (assuming good intent in the DAG to event `origin_server_ts` relationship which can be made up). Maybe we ask another homeserver and see if we get the same answer. We could look forward and backfill to try to close the gap, etc. Maybe ideally, the federation `/timestamp_to_event` endpoint returns events on both sides of the timestamp to reconcile the gap in less steps.

File: synapse/handlers/room.py [hunk 7]
```diff
@@ -1479,25 +1480,41 @@ async def get_event_for_timestamp(
...
                     )
 
-                    # TODO: Do we want to persist this as an extremity?
-                    # TODO: I think ideally, we would try to backfill from
-                    # this event and run this whole
-                    # `get_event_for_timestamp` function again to make sure
-                    # they didn't give us an event from their gappy history.
                     remote_event_id = remote_response.event_id
-                    origin_server_ts = remote_response.origin_server_ts
+                    remote_origin_server_ts = remote_response.origin_server_ts
```
Comment (MadLittleMods) [inline synapse/handlers/room.py:1484]: Since we're backfilling the event, we don't have to rely on the unverifiable `origin_server_ts` that the server tells us anymore. 

TODO: Use the `origin_server_ts` from the PDU itself

File: synapse/handlers/federation_event.py [hunk 8]
```diff
@@ -1315,6 +1317,61 @@ async def _handle_marker_event(self, origin: str, marker_event: EventBase) -> No
...
+        event_non_outlier = make_event_from_dict(
+            event_from_response.get_pdu_json(),
+            event_from_response.room_version,
+            internal_metadata_dict=None,
+        )
+        assert not event_non_outlier.internal_metadata.outlier
+
+        # Persist the event we just fetched, including pulling all of the state
+        # and auth events to de-outlier it. This function is weird and
+        # de-outliers but only works on non-outlier events. This also sets up
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: I think we figured this out in https://github.com/matrix-org/synapse/pull/13205#discussion_r918365746

But it's weird without the context. We can de-outlier an event marked as an outlier, it's just that we chose not to and the reason was unclear before.

File: synapse/handlers/federation_event.py [hunk 9]
```diff
@@ -1315,6 +1317,61 @@ async def _handle_marker_event(self, origin: str, marker_event: EventBase) -> No
...
+                "Unable to find event_id=%s from destination=%s to backfill."
+                % (event_id, destination),
+                affected=event_id,
+            )
+
+        # We want to make a non-outlier event so it plays well with
+        # `_process_pulled_events()` -> `_update_outliers_txn()` to create a
+        # `state_group` and mimics what would happen in a regular backfill.
+        # `get_pdu()` can potentially return an `outlier` depending on the cache
+        # which we don't want.
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: No longer relevant since we fixed `get_pdu` to not return `outliers` anymore.

File: synapse/handlers/room.py (hunk 7 above)
Comment (MadLittleMods) [reply synapse/handlers/room.py:1484]: Updated ✅ https://github.com/matrix-org/synapse/pull/13205/commits/b2be2bcdc0ab3d99c7cce91da90eda347f1abd97

File: synapse/handlers/federation_event.py [hunk 10]
```diff
@@ -766,15 +766,22 @@ async def _process_pulled_event(
         """
         logger.info("Processing pulled event %s", event)
```
Comment (richvdh) [inline synapse/handlers/federation_event.py:768]: this line has got lost.

File: synapse/handlers/federation_event.py [hunk 11]
```diff
@@ -766,15 +766,22 @@ async def _process_pulled_event(
...
 
-        # TODO: Why does this matter? The whole point of this function is to
-        # persist random PDU's from backfill. It shouldn't matter whether we saw
-        # them somewhere else first as an outlier, then during backfill. This
-        # function handles de-outliering anyway.
+        # This function should only be used to backfill events. If you're trying
```
Comment (richvdh) [inline synapse/handlers/federation_event.py:769]: > This function should only be used to backfill events

I'm not sure this is either true or particularly helpful. Notably, it rather raises the question of "wtf is the `backfilled` parameter for?". As well as "backfilled" events, it is used to process events returned from `get_missing_events`, which are treated in the much the same way as those *pushed* via a federation `/send`.

I can't really think of a more helpful thing to say "This function should not be used to persist outliers".

File: synapse/handlers/federation_event.py [hunk 12]
```diff
@@ -766,15 +766,22 @@ async def _process_pulled_event(
...
-        ), "pulled event unexpectedly flagged as outlier"
+        # `EventBase` is used to represent both an event we have not yet
+        # persisted, and one that we have persisted and now keep in the cache.
+        # In an ideal world this method would only be called with the first type
+        # of event, but it turns out that's not actually the case and for
+        # example, you could get an event from cache that is marked as an
+        # `outlier` (fix up that spot though).
+        assert not event.internal_metadata.is_outlier(), (
+            "This is a safe-guard to make sure you're not trying to persist an outlier using this function (use something else). "
+            "If you're trying to backfill an event, this is the right method but you need pass in an event copy that doesn't have `event.internal_metada.outlier = true`."
```
Comment (richvdh) [inline synapse/handlers/federation_event.py:783]: bearing in mind this is going to pop up in a user's logs if/when we mess up, I'm not sure this level of detail is that helpful. By all means include these details in a comment, but I'd keep the assertion message short and sweet ("outlier event passed to `_process_pulled_event`" or something)

File: synapse/handlers/federation_event.py [hunk 13]
```diff
@@ -1314,19 +1324,22 @@ async def _handle_marker_event(self, origin: str, marker_event: EventBase) -> No
...
             destination: The homeserver to pull the given event_id from.
             room_id: The room where the event is from.
             event_id: The event ID to backfill.
+
+        Raises:
+            FederationError if we are unable to find the event from the destination
         """
         logger.info(
-            "backfill_event event_id=%s from destination=%s", event_id, destination
+            "backfill_event_id: event_id=%s from destination=%s", event_id, destination
```
Comment (richvdh) [inline synapse/handlers/federation_event.py:1342]: this has got mangled I think

**Review by richvdh — CHANGES_REQUESTED** [review]
I'm starting to find this rather hard to follow. Could you separate the updates to `get_pdu` and the assertions from the MSC3030-related changes?

File: synapse/handlers/federation_event.py (hunk 12 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:783]: Comment moved to the other PR: https://github.com/matrix-org/synapse/pull/13320#discussion_r923903077

File: synapse/handlers/federation_event.py (hunk 11 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py:769]: Updated in https://github.com/matrix-org/synapse/pull/13320

File: synapse/handlers/federation_event.py (hunk 3 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: `get_pdu` changes split out to https://github.com/matrix-org/synapse/pull/13320

File: synapse/handlers/federation_event.py (hunk 13 above)
Comment (MadLittleMods) [reply synapse/handlers/federation_event.py]: It is outdated now but still don't know what this was about.

File: synapse/handlers/room.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/handlers/room.py:1494]: right, yup, I'm with you now, but I don't think the comment explains this very clearly - or at least, it's not the terminology I would use.

How about:

```
                    # TODO: the requested timestamp may lie in a part of
                    #   the event graph that the remote server *also* doesn't have,
                    #   in which case they will have returned another event which
                    #   may be nowhere near the requested timestamp.
                    #   In future, we may need to ask other homeservers, and/or
                    #   extend `/timestamp_to_event` to return events on *both*
                    #   sides of the timestamp to help reconcile the gap.
```

(by the way, note that there is a loose convention that the second and subsequent lines of a multiline FIXME or TODO should be indented, though I don't know if this is honoured by anything other than the JetBrains IDEs).

Extensions to `/timestamp_to_event` certainly seem like they are going to be needed here.

File: synapse/handlers/federation_event.py (hunk 13 above)
Comment (richvdh) [reply synapse/handlers/federation_event.py]: yup, no idea what was going on here. I'm going to blame github for showing me a mangled diff.

**Review by richvdh — APPROVED** [review]
This is much clearer now that #13320 is factored out, thanks!

lgtm other than the below

File: synapse/handlers/room.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/room.py]: Updated with a slightly modified version of that 👍

**Comment by MadLittleMods** [comment]
@richvdh Thanks for the review and prompt to split things up! 🦏
